src/Foundatio/Messaging/MessageBusBase.cs

```diff
@@ -190,19 +190,23 @@ public abstract class MessageBusBase<TOptions> : IMessageBus, IDisposable where
                     if (isTraceLogLevelEnabled)
                         _logger.LogTrace("Calling subscriber action: {SubscriberId}", subscriber.Id);
 
-                    try {
-                        if (subscriber.Type == typeof(IMessage))
-                            await subscriber.Action(message, subscriber.CancellationToken).AnyContext();
-                        else
-                            await subscriber.Action(body.Value, subscriber.CancellationToken).AnyContext();
-                        
-                        if (isTraceLogLevelEnabled)
-                            _logger.LogTrace("Finished calling subscriber action: {SubscriberId}", subscriber.Id);
-                    } catch (Exception ex) {
-                        if (_logger.IsEnabled(LogLevel.Warning))
-                            _logger.LogWarning(ex, "Error sending message to subscriber: {ErrorMessage}", ex.Message);
-                    }
+                    if (subscriber.Type == typeof(IMessage))
+                        await subscriber.Action(message, subscriber.CancellationToken).AnyContext();
+                    else
+                        await subscriber.Action(body.Value, subscriber.CancellationToken).AnyContext();
+
+                    if (isTraceLogLevelEnabled)
+                        _logger.LogTrace("Finished calling subscriber action: {SubscriberId}", subscriber.Id);
                 });
+            });
+
+            try {
+                await Task.WhenAll(subscriberHandlers.ToArray());
+            } catch (Exception ex) {
+                if (_logger.IsEnabled(LogLevel.Warning))
+                    _logger.LogWarning(ex, "Error sending message to subscribers: {ErrorMessage}", ex.Message);
+
+                throw;
```

Comment:
Can we move this up, and then just return the Task.WhenAll and keep this async. This would ensure we log all exceptions inside of SendMessageToSubscribers

Reply:
Not sure I follow what you mean here.

As far as I can tell, the logging now is almost the same as it was before the change except for the fact that instead of getting a distinct log message for each subscriber handler action that throws an exception there will be a single log message aggregating (AggregateException) all of the exceptions.  Also because the aggregate exception is thrown the trace log message at the bottom doesn't happen either which it would have before.

The swallowing of the exceptions is really just an implementation concern of `InMemoryMessageBus` I think.  Other implementations would be unlikely to be sending messages in the same method that publishes them as they'd be in separate processes (i.e. not in memory) wouldn't they?

Perhaps inside the `InMemoryMessageBus` exception handler that swallows the exception we should just be logging a message as well?

Reply:
I'm just thinking that we could be eating exceptions in this method and only capturing the exceptions the tasks themselves throw. Moving the logging up to my previous comment located to the empty catch block here -> `await SendMessageToSubscribers(messageData);` would ensure all the logic to setting up the tasks would be covered by logging (lines 153-179).

> The swallowing of the exceptions is really just an implementation concern of InMemoryMessageBus I think. Other implementations would be unlikely to be sending messages in the same method that publishes them as they'd be in separate processes (i.e. not in memory) wouldn't they?

Generally yes.

Reply:
Right I think I see now. To clarify, if the call to `SendMessageToSubscribers` in `InMemoryMessageBus` fails for any reason other than a subscriber handler exception we will both swallow it and there'll be no log message?

In which case would it be best to keep the logging in the base class as it is and have the `InMemoryMessageBus` log the exception?  Again, having not used it, and without knowing too much about the use case of the `InMemoryMessageBus` it does seem like an implementation specific concern.  I'd have thought it'd be better to let the exception propagate under normal circumstances (i.e. not swallow it at all) but in order to pass the tests I had to do something with it to allow the `CanTolerateSubscriberFailureAsync` to pass for the `InMemoryMessageBus`.

Anyway I'll put a log message in and see what you think.

Reply:
Sorry for all the back and forth on this, we'll get this merged ASAP. I was wondering this morning should we be eating exceptions when you call publish in the in memory message bus. Seems like you'd want to know if an error occurred during publish, perhaps log and rethrow or just don't even try catch it? Guess it's nice to have logging.